src/porepy/numerics/ad/operator_functions.py

```diff
@@ -0,0 +1,463 @@
+        Arguments are expected to be of type
+        :class:`~porepy.ad.Variable` or :class:`~porepy.ad.MergedVariable`.
+
+        The children are evaluated and respective AD arrays are passed to the abstract
+        `get_values` and `get_jacobian` methods.
+        """
+        children = [self, *args]
+        op = Operator(tree=Tree(self._operation, children=children))
+        return op
+
+    def __repr__(self) -> str:
+        """String representation of this operator function.
+        Uses currently only the given name."""
+
+        s = f"AD Operator function with name {self._name}"
+        return s
+
+    def __mul__(self, other):
+        raise RuntimeError(
+            "ad.Operator functions should only be called, not multiplied."
+        )
+
+    def __add__(self, other):
+        raise RuntimeError("ad.Operator functions should only be called, not added.")
+
+    def __sub__(self, other):
+        raise RuntimeError(
+            "ad.Operator functions should only be called, not subtracted."
+        )
+
+    def __div__(self, other):
+        raise RuntimeError("ad.Operator functions should only be called, not divided.")
+
+    def __truediv__(self, other):
+        raise RuntimeError("ad.Operator functions should only be called, not divided.")
+
+    def parse(self, gb: "pp.GridBucket"):
+        """Parsing to an numerical value.
+
+        The real work will be done by combining the function with arguments, during
+        parsing of an operator tree.
+
+        Parameters:
+        :param gb: Mixed-dimensional grid. Not used, but it is needed as input to be
+            compatible with parse methods for other operators.
+        :type gb: :class:`~porepy.GridBucket`
+
+        :return: the instance itself
+        """
+        return self
+
+    @abc.abstractmethod
+    def get_values(self, *args: "Ad_array") -> "np.ndarray":
+        """
+        Abstract method for evaluating the callable passed at instantiation.
+
+        The AD arrays passed as arguments will be in the same order as the operators passed to
+        the call to this instance.
+
+        The returned numpy array will be be set as 'val' argument for the AD array representing
+        this instance.
+        """
+        pass
+
+    @abc.abstractmethod
+    def get_jacobian(self, *args: "Ad_array") -> "sps.spmatrix":
+        """
+        Abstract method for evaluating the Jacobian of the callable passed at instantiation.
+
+        The AD arrays passed as arguments will be in the same order as the operators passed to
+        the call to this instance.
+
+        The returned numpy array will be be set as 'jac' argument for the AD array representing
+        this instance.
+
+        NOTE: the necessary dimensions for the jacobian can be extracted from the dimensions
+        of the jacobian of passed Ad_array instances.
+        """
+        pass
+
+
+class AbstractJacFunction(AbstractFunction):
```

Comment:
I'm not sure about the naming convention here, we have been moving towards full names (so, Jacobian rather than Jac).